source/logrepl/snapshot.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+)
+
+var psql = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+
+const actionSnapshot string = "snapshot"
+
+type SnapshotConfig struct {
+	URI          string
+	SnapshotName string
+	Table        string
+	Columns      []string
+	KeyColumn    string
+}
+
+type SnapshotIterator struct {
+	config SnapshotConfig
+
+	conn *pgx.Conn
+	rows pgx.Rows
+
+	internalPos int64
+}
+
+func NewSnapshotIterator(ctx context.Context, conn *pgx.Conn, cfg SnapshotConfig) (*SnapshotIterator, error) {
+	s := &SnapshotIterator{
+		config: cfg,
+		conn:   conn,
+	}
+
+	err := s.startSnapshotTx(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start snapshot tx: %w", err)
+	}
+
+	err = s.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load rows: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *SnapshotIterator) loadRows(ctx context.Context) error {
+	query, args, err := psql.
+		Select(s.config.Columns...).
+		From(s.config.Table).
+		ToSql()
+	if err != nil {
+		return fmt.Errorf("failed to create read query: %w", err)
+	}
+
+	rows, err := s.conn.Query(ctx, query, args...)
+	if err != nil {
+		return fmt.Errorf("failed to query rows: %w", err)
+	}
+	s.rows = rows
+
+	return nil
+}
+
+func (s *SnapshotIterator) startSnapshotTx(ctx context.Context) error {
+	_, err := s.conn.Exec(ctx, `BEGIN ISOLATION LEVEL REPEATABLE READ;`)
+	if err != nil {
+		return nil
+	}
+
+	snapshotTx := fmt.Sprintf(`SET TRANSACTION SNAPSHOT '%s'`, s.config.SnapshotName)
+	_, err = s.conn.Exec(ctx, snapshotTx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	if !s.rows.Next() {
+		if err := s.rows.Err(); err != nil {
+			return sdk.Record{}, fmt.Errorf("rows error: %w", err)
+		}
+
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	return s.buildRecord(ctx)
+}
+
+// Ack is a noop for snapshots
+func (s *SnapshotIterator) Ack(ctx context.Context, pos sdk.Position) error {
+	return nil // noop for snapshots
+}
+
+// Teardown attempts to gracefully teardown the iterator.
+func (s *SnapshotIterator) Teardown(ctx context.Context) error {
+	s.rows.Close()
+	return s.commit(ctx)
+}
+
+func (s *SnapshotIterator) commit(ctx context.Context) error {
+	_, err := s.conn.Exec(ctx, `COMMIT;`)
+	if err != nil {
+		return fmt.Errorf("failed to commit: %w", err)
+	}
+	return nil
+}
+
+func (s *SnapshotIterator) buildRecord(ctx context.Context) (sdk.Record, error) {
+	if err := s.rows.Err(); err != nil {
+		return sdk.Record{}, fmt.Errorf("build record rows error: %w", err)
+	}
+
+	r, err := withPayloadAndKey(sdk.Record{}, s.rows, s.config.Columns, s.config.KeyColumn)
+	if err != nil {
+		return sdk.Record{}, err
+	}
+
+	r.CreatedAt = time.Now()
+
+	r.Metadata = map[string]string{
+		"action": actionSnapshot,
+		"table":  s.config.Table,
+	}
+
+	r.Position = sdk.Position(strconv.FormatInt(s.internalPos, 10))
```

Comment:
I'm thinking how we could make positions a bit more unique. We don't want to hit [this error](https://github.com/ConduitIO/conduit/blob/77abccad616f11b89e154292c7ea37080d17f4bb/pkg/pipeline/stream/acker.go#L163) when we run two postgres source connectors in the same pipeline. We could at least prefix the position with the table name.

Reply:
I decided to use table names for now, if we need to increase specificity in the future we can add the database name to the string fairly easily.